### ISSUE TITLE
Boolean mask setitem with differentiable tensors fails in dynamo

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#102841]
### 🐛 Describe the bug

```
    def test_setitem_boolean_mask_diff(self):
        def fn(x, b, y):
            x = x.clone()
            x[b] = y
            return x

        opt_fn = torch._dynamo.optimize("aot_eager")(fn)
        x = torch.randn(4, requires_grad=True)
        b = torch.tensor([True, False, True, False])
        y = torch.randn(2, requires_grad=True)
        opt_fn(x, b, y)
```

This test fails with

```
  File "/data/users/ezyang/b/pytorch/torch/_meta_registrations.py", line 1548, in meta_index_Tensor
    nonzero = index.nonzero()
  File "/data/users/ezyang/b/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/data/users/ezyang/b/pytorch/torch/_subclasses/fake_tensor.py", line 1153, in __torch_dispatch__       
    return self.dispatch(func, types, args, kwargs)
  File "/data/users/ezyang/b/pytorch/torch/_subclasses/fake_tensor.py", line 1373, in dispatch
    op_impl_out = op_impl(self, func, *args, **kwargs)
  File "/data/users/ezyang/b/pytorch/torch/_subclasses/fake_tensor.py", line 488, in nonzero
    raise DynamicOutputShapeException(func)
torch._dynamo.exc.BackendCompilerFailed: backend='compiler_fn' raised:
DynamicOutputShapeException: aten.nonzero.default
```

### Versions

main

cc @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy

**Comment by ezyang** [comment]
With `@torch._dynamo.config.patch(dynamic_shapes=True, capture_dynamic_output_shape_ops=True)`

```
  File "/data/users/ezyang/b/pytorch/torch/autograd/__init__.py", line 319, in grad
    result = Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File "/data/users/ezyang/b/pytorch/torch/fx/experimental/symbolic_shapes.py", line 814, in guard_bool
    r = self.shape_env.evaluate_expr(self.expr, self.hint)
  File "/data/users/ezyang/b/pytorch/torch/fx/experimental/symbolic_shapes.py", line 2836, in evaluate_expr
    raise self._make_data_dependent_error(expr.xreplace(self.var_to_val), expr)
torch._dynamo.exc.BackendCompilerFailed: backend='compiler_fn' raised:
GuardOnDataDependentSymNode: It appears that you're trying to get a value out of symbolic int/float whose value is data-dependent (and thus we do not know the true value.)  The expression we were trying to evaluate is Eq(i0, 2) (unhinted: Eq(i0, 2)).  Scroll up to see where each of these data-dependent accesses originally occurred.
```

Which, I guess is expected, but also irritating because this situation is realistic when you had a graph break and now we no longer know there's a relationship between the boolean mask and the elements to write it...

**Comment by ezyang** [comment]
The right thing to do is to force the unification because we know that we cannot fail this test, and then add a guard / runtime test as necessary. But this requires us to know when guards are guarding over error, which we don't have infra for right now.
